c++/src/kj/compat/http.c++

```diff
@@ -3489,6 +3489,338 @@ WebSocketPipe newWebSocketPipe() {
+  //  - The text before the delimiter is returned.
+  // If no delimiter is found:
+  //  - `input` is updated to an empty string.
+  //  - The text that had been in `input` is returned.
+  //
+  // (It's up to the caller to stop the loop once `input` is empty.)
+  KJ_IF_MAYBE(index, cursor.findFirst(delimiter)) {
+    auto part = cursor.slice(0, *index);
+    cursor = cursor.slice(*index + 1);
+    return part;
+  }
+  auto result = cursor.asArray();
+  cursor = nullptr;
+
+  return result;
+}
+
+void stripLeadingSpace(ArrayPtr<const char>& str) {
+  // Removes any leading spaces from `str`, modifying it in-place.
+  while (str.size() > 0 && (str[0] == ' ' || str[0] == '\t')) {
+    str = str.slice(1, str.size());
+  }
+}
+
+kj::Vector<kj::String> splitPartsWithCondition(StringPtr cursor, char delim, StringPtr condition) {
+  // Given a string `input` and a delimiter `delim`, split the string into a vector of substrings,
+  // separated by the delimiter. Note that leading whitespace is stripped from each element.
+  //
+  // Optionally, the caller may pass a `condition` string, requesting that each substring starts
+  // with the value inside `condition`. This allows for filtering in 1 pass, rather than 2.
+  kj::Vector<kj::String> parts;
+
+  while (cursor.size() != 0) {
+    auto part = splitNext(cursor, delim);
+    stripLeadingSpace(part);
+
+    if (part.startsWith(condition.asArray())) {
+      parts.add(kj::str(part));
+    }
+  }
+
+  return parts;
+}
+
+kj::Array<KeyMaybeVal> toKeysAndVals(const kj::ArrayPtr<kj::String>& params) {
+  // Given a collection of parameters (a single offer), parse the parameters into <key, MaybeValue>
+  // pairs. If the parameter contains an `=`, we set the `key` to everything before, and the `value`
+  // to everything after. Otherwise, we set the `key` to be the entire parameter.
+  auto result = kj::heapArray<KeyMaybeVal>(params.size());
+  size_t count = 0;
+  for (const auto& param : params) {
+    kj::ArrayPtr<const char> key;
+    kj::Maybe<kj::ArrayPtr<const char>> value;
+
+    KJ_IF_MAYBE(index, param.findFirst('=')) {
+      // Found '=' so we have a value.
+      key = param.slice(0, *index);
+      value = param.slice(*index + 1, param.size());
+    } else {
+      key = param.asArray();
+    }
+
+    result[count].key = key;
+    result[count].val = value;
+    ++count;
+  }
+  return kj::mv(result);
+}
+
+struct ParamType {
+  enum { CLIENT, SERVER } side;
+  enum { NO_CONTEXT_TAKEOVER, MAX_WINDOW_BITS } property;
+};
+
+inline kj::Maybe<ParamType> parseKeyName(kj::ArrayPtr<const char>& key) {
+  // Returns a `ParamType` struct if the `key` is valid and nullptr if invalid.
+  //
+  // The following `key`s are valid, assume x = "client" or "server".
+  //  - x_no_context_takeover
+  //  - x_max_window_bits
+  // Trailing whitespace is allowed -- any other character invalidates the `key`.
+  size_t keySize = 0;
+  ParamType paramType;
+
+  if (key.startsWith("client_no_context_takeover"_kj)) {
```

Comment:
Why `startsWith` and not `==`?

Reply:
I guess because you're checking for trailing whitespace below, but couldn't you trip the trailing whitespace _before_ comparing against all the names? That seems more intuitive.

Reply:
You don't need `findFirst()`. You can implement stripping whitespace from the back of the array the same way you implement stripping from the front:

```
while (str.size() > 0 && (str.back() == ' ' || str.back() == '\t')) {
  str = str.slice(0, str.size() - 1);
}
```

Reply:
https://github.com/capnproto/capnproto/pull/1480/commits/b7bfdca9454aad3e6b78367b1ba3186f385622bc